File: packages/types/src/dynamo/finding.ts
import type {
  AgentType,
  FindingId,
  ISODateString,
  RepoId,
  ReportId,
  Severity,
} from '../common.js';

export type FindingCategory =
  | 'security'
  | 'quality'
  | 'performance'
  | 'cost'
  | 'compliance'
  | 'style'
  | 'infrastructure'
  | 'dependency';

export interface FindingLocation {
  readonly file: string;
  readonly startLine: number;
  readonly endLine?: number;
  readonly column?: number;
}

export interface FindingItem {
  readonly PK: `FINDING#${FindingId}`;
  readonly SK: 'METADATA';
  readonly findingId: FindingId;
  readonly reportId: ReportId;
  readonly repoId: RepoId;
  readonly agent: AgentType;
  readonly severity: Severity;
  readonly category: FindingCategory;
  readonly title: string;
  readonly description: string;
  readonly evidence: string;
  readonly location?: FindingLocation;
  readonly fixable: boolean;
  readonly confidence: number;        // 0.0 - 1.0
  readonly fixSuggestion?: string;
  readonly ruleId?: string;
  readonly references?: readonly string[];
  readonly createdAt: ISODateString;
  // GSI1: reportId + severity
  readonly GSI1PK: `REPORT#${ReportId}`;
  readonly GSI1SK: `${Severity}#${ISODateString}`;
  // GSI2: repoId#category + createdAt
  readonly GSI2PK: `REPO#${RepoId}#${FindingCategory}`;
  readonly GSI2SK: ISODateString;
}

export interface Finding
  extends Omit<FindingItem, 'PK' | 'SK' | 'GSI1PK' | 'GSI1SK' | 'GSI2PK' | 'GSI2SK'> {}

/** Canonical Finding payload emitted by agents (before DynamoDB storage). */
export interface AgentFinding {
  readonly agent: AgentType;
  readonly severity: Severity;
  readonly category: FindingCategory;
  readonly title: string;
  readonly description: string;
  readonly evidence: string;
  readonly location?: FindingLocation;
  readonly fixable: boolean;
  readonly confidence: number;
  readonly fixSuggestion?: string;
  readonly ruleId?: string;
  readonly references?: readonly string[];
}

File: libs/finding-schema/src/finding_schema/finding.py
from __future__ import annotations

from enum import StrEnum
from typing import Annotated

from pydantic import BaseModel, Field, field_validator


class Severity(StrEnum):
    CRITICAL = "CRITICAL"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"
    INFO = "INFO"


class FindingCategory(StrEnum):
    SECURITY = "security"
    QUALITY = "quality"
    PERFORMANCE = "performance"
    COST = "cost"
    COMPLIANCE = "compliance"
    STYLE = "style"
    INFRASTRUCTURE = "infrastructure"
    DEPENDENCY = "dependency"


class FindingLocation(BaseModel):
    model_config = {"frozen": True}

    file: str
    start_line: int
    end_line: int | None = None
    column: int | None = None


class AgentFinding(BaseModel):
    """Canonical finding payload emitted by agents. Stored in DynamoDB + S3."""

    model_config = {"frozen": True}

    agent: str  # AgentType literal
    severity: Severity
    category: FindingCategory
    title: str
    description: str
    evidence: str
    location: FindingLocation | None = None
    fixable: bool = False
    confidence: Annotated[float, Field(ge=0.0, le=1.0)] = 0.8
    fix_suggestion: str | None = None
    rule_id: str | None = None
    references: list[str] = Field(default_factory=list)

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("title must not be empty")
        return v.strip()

    @field_validator("confidence")
    @classmethod
    def confidence_range(cls, v: float) -> float:
        if not 0.0 <= v <= 1.0:
            raise ValueError("confidence must be between 0 and 1")
        return round(v, 4)
